File: 11_ContainerWithMostWater.js
/**
 * @param {number[]} height
 * @return {number}
 */

//解法1 跑巢狀雙層迴圈 計算每一個可能的容積 此方法會超時
const maxArea = function (height) {
    //最大體積(結果)
    let maxVoulume = 0;
    //跑迴圈計算最大體積
    for (let i = 0; i < height.length; i++) {

        for (let j = i + 1; j < height.length; j++) {
            maxVoulume = Math.max(maxVoulume, Math.min(height[i], height[j]) * (j - i))
        }
    }
    return maxVoulume;
};


//解法2 從頭尾開始計算容積  逐漸向內縮小 且須比較左右兩端點高度 較低的向內縮進
const maxArea2 = function (height) {
    //最大體積(結果)
    let maxVoulume = 0;
    //頭尾的索引值
    let left = 0;
    let right = height.length - 1;
    //迴圈條件 right > left
    while (right > left) {
        maxVoulume = Math.max(maxVoulume, Math.min(height[left], height[right]) * (right - left));
        if (height[right] > height[left]) {
            left++;
        } else {
            right--;
        }
    }
    return maxVoulume;
}

File: a11.ContainerWithMostWater/11_ContainerWithMostWater.py
# 解法1  雙指針  從左右向內縮小並比較左右兩端點高度
# 較低的向內縮進 相等時左端右移
def maxArea(height: list[int]) -> int:
    # 最大容積
    maxVoulume = 0
    # 頭尾索引
    left = 0
    right = len(height) - 1
    # 迴圈進行計算
    while right > left:
        # 比對當前計算容積與maxVolume誰大
        maxVoulume = max(maxVoulume, min(
            height[left], height[right]) * (right - left))
        if height[left] <= height[right]:
            left += 1
        else:
            right -= 1
    return maxVoulume


maxArea([10, 14, 10, 4, 10, 2, 6, 1, 6, 12])
